NodeApp.render

### Astro Info

```block
Astro                    v5.14.6
Vite                     v6.4.0
Node                     v22.20.0
System                   macOS (arm64)
Package Manager          npm
Output                   static
Adapter                  @astrojs/node (v9.5.0)
Integrations             none
```

### If this issue only occurs in one browser, which browser is a problem?

_No response_

### Describe the Bug

I ran a benchmark using `wrk` on `npx astro build && npx astro preview`.

Stress testing `http://localhost:4321` works fine with no issues.

However, when I append `http://localhost:4321/#`, an error is logged, likely due to a Node adapter bug.


```

5:59:48 [ERROR] TypeError: mod.page is not a function
    at NodeApp.render (file:///Users/s/starter/dist/server/chunks/_@astrojs-ssr-adapter_D-0bE0Wa.mjs:3495:55)
    at async file:///Users/s/starter/dist/server/chunks/_@astrojs-ssr-adapter_D-0bE0Wa.mjs:3971:24
15:59:48 [ERROR] TypeError: mod.page is not a function
    at NodeApp.render (file:///Users/s/starter/dist/server/chunks/_@astrojs-ssr-adapter_D-0bE0Wa.mjs:3495:55)
    at async file:///Users/s/starter/dist/server/chunks/_@astrojs-ssr-adapter_D-0bE0Wa.mjs:3971:24
15:59:48 [ERROR] TypeError: mod.page is not a function
    at NodeApp.render (file:///Users/s/starter/dist/server/chunks/_@astrojs-ssr-adapter_D-0bE0Wa.mjs:3495:55)
    at async file:///Users/s/starter/dist/server/chunks/_@astrojs-ssr-adapter_D-0bE0Wa.mjs:3971:24
15:59:48 [ERROR] TypeError: mod.page is not a function
    at NodeApp.render (file:///Users/s/starter/dist/server/chunks/_@astrojs-ssr-adapter_D-0bE0Wa.mjs:3495:55)
    at async file:/
```

### What's the expected result?

Should handle `/#`.

https://github.com/wildfiremedia/astrossr

NA

### Participation

- [ ] I am willing to submit a pull request for this issue.

--- a/.changeset/lazy-jeans-wave.md
+++ b/.changeset/lazy-jeans-wave.md
@@ -1,0 +1,5 @@
+---
+'@astrojs/node': patch
+---
+
+Fixes an issue where some prendered pages weren't correctly rendered when using the Node.js adapter in middleware mode.

--- a/packages/astro/src/core/app/index.ts
+++ b/packages/astro/src/core/app/index.ts
@@ -552,6 +552,7 @@ export class App {
 			session = renderContext.session;
 			response = await renderContext.render(await mod.page());
 		} catch (err: any) {
+			this.#logger.error('router', "Error while trying to render the route " + routeData.route)
 			this.#logger.error(null, err.stack || err.message || String(err));
 			return this.#renderError(request, {
 				locals,

--- a/packages/integrations/node/package.json
+++ b/packages/integrations/node/package.json
@@ -49,6 +49,9 @@
     "cheerio": "1.1.2",
     "devalue": "^5.6.1",
     "express": "^4.22.1",
+		"fastify":"^5.6.2",
+		"@fastify/middie": "^9.1.0",
+		"@fastify/static":"^9.0.0",
     "node-mocks-http": "^1.17.2"
   },
   "publishConfig": {

--- a/packages/integrations/node/src/serve-static.ts
+++ b/packages/integrations/node/src/serve-static.ts
@@ -20,7 +20,13 @@ export function createStaticHandler(app: NodeApp, options: Options) {
 	 */
 	return (req: IncomingMessage, res: ServerResponse, ssr: () => unknown) => {
 		if (req.url) {
-			const [urlPath, urlQuery] = req.url.split('?');
+			// There might be cases where the incoming URL has the #, which we want to remove.
+			let fullUrl = req.url;
+			if (req.url.includes("#")) {
+				fullUrl = fullUrl.slice(0, req.url.indexOf("#"))
+			} 
+			
+			const [urlPath, urlQuery] = fullUrl.split('?');
 			const filePath = path.join(client, app.removeBase(urlPath));
 
 			let isDirectory = false;

--- a/packages/integrations/node/test/fixtures/node-middleware/src/pages/dyn/[dynamic].astro
+++ b/packages/integrations/node/test/fixtures/node-middleware/src/pages/dyn/[dynamic].astro
@@ -1,0 +1,26 @@
+---
+export const prerender = false;
+
+interface Props {
+	dynamic: string
+}
+
+export async function getStaticPaths() {
+	return [
+		{ params: { dynamic: 'foo' } },
+		{ params: { dynamic: 'bar' } }
+	];
+}
+
+const { dynamic } = Astro.params;
+
+---
+
+<html lang="en">
+<head><title>node-middleware</title></head>
+<style>
+</style>
+<body>
+<div>{dynamic}</div>
+</body>
+</html>

--- a/packages/integrations/node/test/node-middleware.test.js
+++ b/packages/integrations/node/test/node-middleware.test.js
@@ -4,6 +4,10 @@ import * as cheerio from 'cheerio';
 import express from 'express';
 import nodejs from '../dist/index.js';
 import { loadFixture, waitServerListen } from './test-utils.js';
+import Fastify from 'fastify';
+import fastifyMiddie from '@fastify/middie';
+import fastifyStatic from '@fastify/static';
+import { fileURLToPath } from 'node:url';
 
 /**
  * @typedef {import('../../../astro/test/test-utils').Fixture} Fixture
@@ -50,7 +54,7 @@ describe('behavior from middleware, standalone', () => {
 	});
 });
 
-describe('behavior from middleware, middleware', () => {
+describe('behavior from middleware, middleware with express', () => {
 	/** @type {import('./test-utils').Fixture} */
 	let fixture;
 	let server;
@@ -76,7 +80,7 @@ describe('behavior from middleware, middleware', () => {
 		delete process.env.PRERENDER;
 	});
 
-	it('when mode is middleware', async () => {
+	it('should render the endpoint', async () => {
 		const res = await fetch('http://localhost:8889/ssr');
 
 		assert.equal(res.status, 200);
@@ -86,5 +90,133 @@ describe('behavior from middleware, middleware', () => {
 
 		const body = $('body');
 		assert.equal(body.text().includes("Here's a random number"), true);
+	});
+
+	it('should render the index.html page [static]', async () => {
+		const res = await fetch('http://localhost:8889/');
+
+		assert.equal(res.status, 200);
+
+		const html = await res.text();
+		const $ = cheerio.load(html);
+
+		const body = $('body');
+		assert.equal(body.text().includes("1"), true);
+	});
+
+	it('should render the index.html page [static] when the URL has the hash', async () => {
+		const res = await fetch('http://localhost:8889/#');
+
+		assert.equal(res.status, 200);
+
+		const html = await res.text();
+		const $ = cheerio.load(html);
+
+		const body = $('body');
+		assert.equal(body.text().includes("1"), true);
+	});
+
+	it('should render the dynamic pages', async () => {
+		let res = await fetch('http://localhost:8889/dyn/foo');
+
+		assert.equal(res.status, 200);
+
+		let html = await res.text();
+		let $ = cheerio.load(html);
+
+		let body = $('body');
+		assert.equal(body.text().includes("foo"), true);
+
+		res = await fetch('http://localhost:8889/dyn/bar');
+
+		assert.equal(res.status, 200);
+
+		html = await res.text();
+		$ = cheerio.load(html);
+
+		body = $('body');
+		assert.equal(body.text().includes("bar"), true);
+	});
+});
+
+
+describe('behavior from middleware, middleware with fastify', () => {
+	/** @type {import('./test-utils').Fixture} */
+	let fixture;
+	let server;
+
+	before(async () => {
+		process.env.PRERENDER = false;
+		fixture = await loadFixture({
+			root: './fixtures/node-middleware/',
+			output: 'server',
+			adapter: nodejs({ mode: 'middleware' }),
+		});
+		await fixture.build();
+		const { handler } = await fixture.loadAdapterEntryModule();
+		const app = Fastify({ logger: false });
+		await app
+			.register(fastifyStatic, {
+				root: fileURLToPath(new URL('./dist/client', import.meta.url)),
+			})
+			.register(fastifyMiddie);
+		app.use(handler);
+
+		await app.listen({ port: 8889 });
+		
+		server = app;
+	});
+
+	after(async () => {
+		server.close();
+		await fixture.clean();
+
+		delete process.env.PRERENDER;
+	});
+
+	it('should render the endpoint', async () => {
+		const res = await fetch('http://localhost:8889/ssr');
+
+		assert.equal(res.status, 200);
+
+		const html = await res.text();
+		const $ = cheerio.load(html);
+
+		const body = $('body');
+		assert.equal(body.text().includes("Here's a random number"), true);
+	});
+
+	it('should render the index.html page [static]', async () => {
+		const res = await fetch('http://localhost:8889');
+
+		assert.equal(res.status, 200);
+
+		const html = await res.text();
+		const $ = cheerio.load(html);
+
+		const body = $('body');
+		assert.equal(body.text().includes("1"), true);
+	});
+
+	it('should render the dynamic pages', async () => {
+		let res = await fetch('http://localhost:8889/dyn/foo');
+
+		assert.equal(res.status, 200);
+
+		let html = await res.text();
+		let $ = cheerio.load(html);
+
+		let body = $('body');
+		assert.equal(body.text().includes("foo"), true);
+
+		res = await fetch('http://localhost:8889/dyn/bar');
+
+		assert.equal(res.status, 200);
+
+		html = await res.text();
+		$ = cheerio.load(html);
+
+		body = $('body');
+		assert.equal(body.text().includes("bar"), true);
 	});
 });

